codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddS3Config.java

```diff
@@ -106,6 +113,89 @@ public Model preprocessModel(Model model, TypeScriptSettings settings) {
+                }
+                StructureShape structureShape = model.expectShape(operationShape.getOutputShape(), StructureShape.class);
+
+                Set<Map.Entry<String, MemberShape>> memberEntries = structureShape
+                    .getAllMembers()
+                    .entrySet();
+                StructureShape.Builder structureShapeBuilder = structureShape.toBuilder();
+
+                boolean isTargetingExpires = structureShape
+                    .getAllMembers()
+                    .values()
+                    .stream()
+                    .anyMatch(memberShape -> memberShape.getTarget().equals(expiresShape.getId()));
+
+                if (isTargetingExpires) {
+                    for (Map.Entry<String, MemberShape> entry : memberEntries) {
+                        String memberName = entry.getKey();
+                        MemberShape memberShape = entry.getValue();
+
+                        if (memberShape.getTarget().equals(expiresShape.getId())) {
+                            structureShapeBuilder
+                                .removeMember(memberName)
+                                .addMember(
+                                    memberName,
+                                    expiresTimestampShape.getId(),
+                                    (m) -> {
+                                        m
+                                            .addTrait(new DocumentationTrait("Deprecated in favor of ExpiresString."))
+                                            .addTrait(memberShape.getTrait(HttpHeaderTrait.class).get())
+                                            .addTrait(DeprecatedTrait.builder().build());
+                                    }
+                                )
+                                .addMember(
+                                    "ExpiresString",
+                                    expiresStringShape.getId(),
+                                    (m) -> {
+                                        m
+                                            .addTrait(memberShape.getTrait(DocumentationTrait.class).get())
+                                            .addTrait(new HttpHeaderTrait("ExpiresString"));
+                                    }
+                                );
+                        } else {
+                            // This is done to preserve the member order
+                            // and insert ExpiresString adjacent to Expires.
+                            structureShapeBuilder
+                                .removeMember(memberName)
+                                .addMember(memberName, memberShape.getTarget(), (m) -> {
+                                    m.addTraits(memberShape.getAllTraits().values());
+                                });
+                        }
+                    }
+                    modelBuilder
+                        .addShape(structureShapeBuilder.build());
+                }
+            }
+        }
+
```

Comment:
This code should initially have been in Java anyway. However, at the time I couldn't figure out how to mutate the model, and wrote it in JS.

It turns out the `.addX()` methods on the various builders generally do a replace on a map collection, and it was easier to figure out this time.